src/libpgmoneta/manifest.c

```diff
@@ -0,0 +1,295 @@
+   return 1;
+}
+
+int
+pgmoneta_parse_manifest(char* manifest_path, struct manifest* manifest)
+{
+   FILE* f = NULL;
+   size_t checksum_size = 0;
+   size_t content_size = 0;
+   size_t size = 0;
+   char* json = NULL;
+   char* ptr = NULL;
+   struct cJSON* manifest_json = NULL;
+   struct cJSON* manifest_checksum = NULL;
+   struct cJSON* files = NULL;
+   struct cJSON* file = NULL;
+
+   if (!pgmoneta_exists(manifest_path))
+   {
+      pgmoneta_log_error("Could not find backup manifest: %s", manifest_path);
+      goto error;
+   }
+   size = pgmoneta_get_file_size(manifest_path);
+   json = (char*) malloc(size + 1);
+
+   f = fopen(manifest_path, "r");
+   if (f == NULL)
+   {
+      pgmoneta_log_error("Could not open backup manifest: %s", manifest_path);
+      goto error;
+   }
+
+   fread(json, 1, size, f);
+   json[size] = '\0';
+
+   manifest_json = cJSON_Parse(json);
+   if (manifest_json == NULL)
+   {
+      const char* error_ptr = cJSON_GetErrorPtr();
+      if (error_ptr != NULL)
+      {
+         pgmoneta_log_error("Unable to parse manifest: %s", error_ptr);
+      }
+      goto error;
+   }
+
+   files = cJSON_GetObjectItemCaseSensitive(manifest_json, "Files");
+   manifest_checksum = cJSON_GetObjectItemCaseSensitive(manifest_json, "Manifest-Checksum");
+   if (!cJSON_IsString(manifest_checksum))
+   {
+      pgmoneta_log_error("Unable to parse manifest %s", manifest_path);
+      goto error;
+   }
+   checksum_size = strlen(manifest_checksum->valuestring);
+   manifest->checksum = (char*)malloc(checksum_size + 1);
+   memcpy(manifest->checksum, manifest_checksum->valuestring, checksum_size);
+   manifest->checksum[checksum_size] = '\0';
+
+   struct manifest_file* tmp = NULL;
+   cJSON_ArrayForEach(file, files)
+   {
+      struct manifest_file* manifest_file = NULL;
+      struct cJSON* file_path = cJSON_GetObjectItemCaseSensitive(file, "Path");
+      struct cJSON* file_size = cJSON_GetObjectItemCaseSensitive(file, "Size");
+      struct cJSON* file_checksum = cJSON_GetObjectItemCaseSensitive(file, "Checksum");
+
+      if (!cJSON_IsNumber(file_size) || !cJSON_IsString(file_path) || !cJSON_IsString(file_checksum))
+      {
+         pgmoneta_log_error("Unable to parse manifest %s", manifest_path);
+         goto error;
+      }
+      manifest_file_init(&manifest_file, file_path->valuestring, file_checksum->valuestring, file_size->valuedouble);
+      if (manifest_file == NULL)
+      {
+         pgmoneta_log_error("Unable to allocated space for manifest file info");
+         goto error;
+      }
+      if (tmp == NULL)
+      {
+         tmp = manifest_file;
+         manifest->files = tmp;
+      }
+      else
+      {
+         tmp->next = manifest_file;
+         tmp = tmp->next;
+      }
+   }
+
+   ptr = strstr(json, "\"Manifest-Checksum\"");
+
+   // save preceding lines before the manifest checksum field,
+   // in order to verify the checksum of the manifest itself
+   content_size = ptr - json;
```

Comment:
This seems hacky - isn't there a better way ?

We can't assume that the manifest format to be stable

Reply:
Oh it looks hacky but is actually not, because the checksum of the manifest itself has to be calculated last. So it's always the last one to be added to the manifest json. According to https://www.postgresql.org/docs/current/backup-manifest-toplevel.html: "This key is always present on the last line of the backup manifest file. "

I could also just count backwards until getting '\n', but I don't know how stable the format is, for example is the feilds always separated by new line? Or will there always be a trailing '\n', so searching for this field directly is actually easier and more reliable I think. Let me know your thoughts.